staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,1048 @@
+
+				return OnPaymentReceived
+
+			case invoices.ContractCanceled:
+				// If the invoice was canceled we only log here
+				// since we still need to monitor until the htlc
+				// timed out.
+				log.Warnf("invoice for swap hash %v canceled",
+					f.loopIn.SwapHash)
+
+			default:
+				err = fmt.Errorf("unexpected invoice state %v "+
+					"for swap hash %v canceled",
+					update.State, f.loopIn.SwapHash)
+
+				return f.HandleError(err)
+			}
+
+		case <-eventCtx.ctx.Done():
+			return f.HandleError(eventCtx.ctx.Err())
+		}
+	}
+}
+
+// SweepHtlcTimeoutAction is called if the server published the htlc tx without
+// paying the invoice. We wait for the timeout path to open up and sweep the
+// funds back to us.
+func (f *FSM) SweepHtlcTimeoutAction(eCtx fsm.EventContext) fsm.EventType {
+	eventCtx, ok := eCtx.(EventContext)
+	if !ok {
+		return f.HandleError(fmt.Errorf("unexpected context type %T",
+			eCtx))
+	}
+
+	for {
+		err := f.createAndPublishHtlcTimeoutSweepTx(eventCtx.ctx)
+		if err == nil {
+			break
+		}
+
+		f.Errorf("unable to create and publish htlc timeout sweep "+
+			"tx: %v, retrying in  %v", err, time.Hour.String())
+
+		select {
+		case <-eventCtx.ctx.Done():
+			f.Errorf(eventCtx.ctx.Err().Error())
+
+		default:
+			<-time.After(1 * time.Hour)
+		}
+	}
+
+	return OnHtlcTimeoutSweepPublished
+}
+
+// MonitorHtlcTimeoutSweepAction is called after the htlc timeout sweep tx has
+// been published. We monitor the confirmation of the htlc timeout sweep tx and
+// finalize the deposits once swept.
+func (f *FSM) MonitorHtlcTimeoutSweepAction(eC fsm.EventContext) fsm.EventType {
+	eventCtx, ok := eC.(EventContext)
+	if !ok {
+		return f.HandleError(fmt.Errorf("unexpected context type %T",
+			eC))
+	}
+
+	f.Infof("monitoring htlc timeout sweep tx %v",
+		f.loopIn.HtlcTimeoutSweepTxHash)
+
+	timeoutSweepPkScript, err := txscript.PayToAddrScript(
+		f.loopIn.HtlcTimeoutSweepAddress,
+	)
+	if err != nil {
+		err = fmt.Errorf("unable to convert timeout sweep address to "+
+			"pkscript: %w", err)
+
+		return f.HandleError(err)
+	}
+
+	htlcTimeoutTxidChan, errChan, err := f.cfg.ChainNotifier.RegisterConfirmationsNtfn( //nolint:lll
+		eventCtx.ctx, f.loopIn.HtlcTimeoutSweepTxHash,
+		timeoutSweepPkScript, defaultConfTarget,
+		int32(f.loopIn.InitiationHeight),
+	)
+	if err != nil {
+		err = fmt.Errorf("unable to register to the htlc timeout "+
+			"sweep tx: %w", err)
+
+		return f.HandleError(err)
+	}
+
+	for {
+		select {
+		case err := <-errChan:
+			return f.HandleError(err)
+
+		case conf := <-htlcTimeoutTxidChan:
+			err = f.cfg.DepositManager.TransitionDeposits(
+				eventCtx.ctx, f.loopIn.Deposits,
+				deposit.OnHtlcTimeoutSwept,
+				deposit.HtlcTimeoutSwept,
+			)
+			if err != nil {
+				err = fmt.Errorf("unable to transition the "+
+					"deposits to the htlc timeout swept "+
+					"state: %w", err)
+
+				return f.HandleError(err)
+			}
+
+			f.Infof("htlc timeout sweep tx confirmed at block-"+
```

Comment:
"htlc timeout sweep tx got %d confirmations at block %d", defaultConfTarget, conf.BlockHeight

The confirmation was at block `conf.BlockHeight - defaultConfTarget` IIUC